app/ModelFunctions/AlbumActions/UpdateTakestamps.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+/** @noinspection PhpUndefinedClassInspection */
+
+namespace App\ModelFunctions\AlbumActions;
+
+use App\Album;
+use App\Photo;
+use Illuminate\Database\Eloquent\Collection;
+
+class UpdateTakestamps
+{
+	/**
+	 * TODO: Check if this code is duplicated.
+	 *
+	 * Recursively go through each sub album and build a list of them.
+	 *
+	 * @param Album      $album
+	 * @param Collection $return
+	 *
+	 * @return Collection
+	 */
+	private static function get_all_sub_albums_id(Album $album)
+	{
+		return $album->children->reduce(function ($collect, $_album) {
+			return $collect
+				->concat([$_album->id])
+				->concat(self::get_all_sub_albums_id($_album));
+		}, new Collection());
+	}
+
+	/**
+	 * Go through each sub album and update the minimum and maximum takestamp of the pictures.
+	 * This is expensive and not normally necessary so we only use it
+	 * during migration.
+	 */
+	public function update_min_max_takestamp(Album $album)
+	{
+		$album_list = self::get_all_sub_albums_id($album, [$album->id]);
+
+		$min = Photo::whereIn('album_id', $album_list)->min('takestamp');
+		$max = Photo::whereIn('album_id', $album_list)->max('takestamp');
+		$album->min_takestamp = $min;
+		$album->max_takestamp = $max;
+	}
+
+	/**
+	 * Update album's min_takestamp and max_takestamp based on changes made
+	 * to the album content.  If needed, recursively updates parent album(s).
+	 *
+	 * @param array $takestamps : an array with the takestamps of changed
+	 *                          elements; for albums needs to include both min and max takestamps
+	 *                          (including null elements in the array is safe)
+	 * @param bool  $adding     :     true if adding new content, false if removing
+	 *
+	 * @return bool: true if successful
+	 */
+	public static function update_takestamps(Album $album, array $takestamps, bool $adding)
+	{
+		// Begin by calculating min and max takestamps from the array.
+		// The array may contain null values, which is why we can't use the
+		// built-in min() function for this (it will always return null if
+		// present).  For consistency, we don't use the built-in max()
+		// either.
+		$minTS = $maxTS = null;
+		foreach ($takestamps as $takestamp) {
+			if ($takestamp !== null) {
+				if ($minTS === null || $minTS > $takestamp) {
+					$minTS = $takestamp;
+				}
+				if ($maxTS === null || $maxTS < $takestamp) {
+					$maxTS = $takestamp;
+				}
+			}
+		}
+		if ($minTS === null || $maxTS === null) {
```

Comment:
I can't decide about this one. The logic is sound (if either is null we don't want to mess with it), but if one is null, both *should* be null. Maybe just...
```suggestion
		}
		// If either minTS or maxTS is null, both should be null so we don't need to update anything.
		if ($minTS === null || $maxTS === null) {
```

That's really bad diffing, and I haven't checked the comment length.